cli/lsp/language_server.rs

```diff
@@ -343,8 +351,14 @@ impl Inner {
     Ok(())
   }
 
+  pub fn update_performance(&mut self) {
+    self
+      .performance
+      .do_logging(self.config.settings.internal_debug);
+  }
```

Comment:
We already have a system for conditional logging using the `debug!` and `info!` macros. These can be enabled with the `--log-level=debug` command-line option. Can we not reuse that system for LSP debugging?

Note that the log-level could be adjusted at runtime as well: 
https://docs.rs/log/0.4.14/log/#implementing-a-logger

Reply:
Yeah, I knew that, the problem is a lot of other crates, especially swc, use the same logging, and so setting the logging flag makes it impossible to sort out what is going on.

Reply:
env_logger provides the ability to filter messages based on regex
https://docs.rs/env_logger/0.8.3/env_logger/#filtering-results

Reply:
Ok, I spent some time looking at this.  A few notes:

- the logger can only be set once and has to be global static
- we use env_logger as our global logger to do this, once we `init()` it appears you can't reconfigure it.
- the `RUST_LOG` is an environment variable, which can be passed in at startup, and the filter `debug/mark|measure` would be suitable, but we can only do that when start/restarting the LSP.
- you can use the `.parse_filters()` method on the builder which follows the same format as `RUST_LOG` but again that has to be determined at startup, versus changing dynamically.

There is the `.format()` closure which maybe could do some "filtering" of records based on a flag, but there doesn't appear to be a dynamic way of changing things in a running process, outside of implementing our own logger (which is what is mentioned in the first link).

Reply:
My guess is the Log::enabled method could be used to dynamically turn on and off logging
https://docs.rs/log/0.4.14/log/trait.Log.html#tymethod.enabled

But it seems a little difficult to access... I guess you'd have to wrap the env_logger.

Reply:
Hmmm... yeah, that should work, but like you said would require wrapping. Ok, I will look at that a bit closer. You can generate the logger without installing it for wrapping, that I know.

Reply:
Ok, I reworked it... the biggest "problem" is that we need to share the state of the logging being enabled between the logger and the `lsp::language_server` and we don't have any infrastructure for that, so I chose a way (`Arc<AtomicBool>`) to do that, but am more than willing to take other suggestions.

Also broke all the logging stuff out to a seperate module, as it just seem unnecessary in an already complex `main.rs`.